File: BOJ/5446.cpp
#include <cstdio>
#include <algorithm>
#include <string.h>
#include <stdio.h>
using namespace std;
int cnt;
struct Trie {
	Trie* tree[80];
	bool finish;
	bool del;
	Trie() {
		fill(tree, tree + 80, nullptr);
		finish = false, del = false;
	}
	~Trie() {
		for (int i = 0; i<80; i++) {
			if (tree[i]) delete tree[i];
		}
	}

	void insert(char *key,bool c) {
		if (*key == NULL) {
			finish = c;
			del = c;
			return;
		}
		else {
			int next = *key-'.';
			if (!tree[next]) {
				tree[next] = new Trie;
			}
			del = c;
			tree[next]->insert(key + 1,c);
		}
	}


	void check() {
		if (del) {
			cnt++;
			return;
		}
		else if (finish) {
			cnt++;
		}
		for (int i = 0; i < 80; i++) {
			if (tree[i]) tree[i]->check();
		}
	}
};
int main() {
	int n,m,t;
	scanf(" %d", &t);
	while (t--) {
		scanf(" %d", &n);
			Trie* trie = new Trie();
			cnt = 0;
			for (int i = 0; i < n; i++) { //지워도 됨
				char s[21] = {'\0',};
				scanf(" %s", s);
				trie->insert(s,true);
			}
			scanf(" %d", &m);
			for (int i = 0; i < m; i++) { //지우면 안됨
				char s[21] = { '\0', };
				scanf(" %s", s);
				trie->insert(s, false);
			}
			trie->check();
			printf("%d\n", cnt);
			delete trie;
	}
	return 0;
}

File: BOJ(Java 공부용)/5446.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int cnt = 0;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int t = Integer.parseInt(st.nextToken());
        while(t-- >0) {
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            Trie trie = new Trie();
            cnt = 0;
            // 지워도 되는것
            for(int i=0;i<n;i++) {
                String s = br.readLine();
                insert(s,trie,true);
            }
            st = new StringTokenizer(br.readLine());
            int m = Integer.parseInt(st.nextToken());
            for(int i=0;i<m;i++) {
                String s = br.readLine();
                insert(s,trie,false);
            }
            check(trie);
            bw.write(cnt+"\n");
        }
        bw.flush();
        bw.close();
    }
    private static class Trie {
        private Trie[] letter;
        private boolean finish;
        private boolean del;
        private Trie() {
            this.letter = new Trie[80];
            this.finish = false;
            this.del = false;
        }
    }
    private static void check(Trie trie) {
        if(trie.del) {
            cnt++;
            return;
        }
        else if(trie.finish) cnt++;
        for(int i=0;i<80;i++) {
            if(trie.letter[i] != null) check(trie.letter[i]);
        }
    }
    private static void insert(String s,Trie trie,boolean d) {
        for(int i=0;i<s.length();i++) {
            int c = s.charAt(i)-'.';
            if(trie.letter[c] == null) trie.letter[c] = new Trie();
            trie.del = d;
            trie = trie.letter[c];
        }
        trie.finish = d;
        trie.del = d;
    }
}